File: Shared/Controllers/WatchlistStore.swift
//
//  WatchlistStore.swift
//  Edendale
//
//  SwiftData is the source of truth for Watchlist. Local mutations are saved
//  first and mirrored to TMDB when an account is connected. Pending actions
//  are durable so a failed or interrupted request is retried on the next sync.
//

import CoreData
import Foundation
import SwiftData

protocol WatchlistRemoteClient: Sendable {
    func watchlistItems(_ type: TMDBMediaType) async throws -> [TMDBMediaItem]
    func setWatchlist(_ inWatchlist: Bool, for ref: MediaRef) async throws
}

extension TMDBAccountClient: WatchlistRemoteClient {}

@MainActor
@Observable
final class WatchlistStore {
    private let modelContext: ModelContext
    private let account: any WatchlistRemoteClient
    private let isAuthenticated: () -> Bool

    /// A compact observable index used by detail-page toggle controls.
    private(set) var items: [WatchlistItem] = []
    private(set) var refs: Set<MediaRef> = []
    private(set) var isSyncing = false
    private(set) var lastSyncError: String?

    @ObservationIgnored private var isRunningRemoteWork = false
    @ObservationIgnored private var pendingPushRequested = false
    @ObservationIgnored private var fullSyncRequested = false
    @ObservationIgnored private var fullSyncInProgress = false

    convenience init(modelContext: ModelContext) {
        self.init(
            modelContext: modelContext,
            account: TMDBAccountClient(),
            isAuthenticated: { TMDBUserSession.current != nil },
            legacyContext: Persistence.cloudPersistentContainer.viewContext
        )
    }

    init(
        modelContext: ModelContext,
        account: any WatchlistRemoteClient,
        isAuthenticated: @escaping () -> Bool,
        legacyContext: NSManagedObjectContext?
    ) {
        self.modelContext = modelContext
        self.account = account
        self.isAuthenticated = isAuthenticated
        migrateLegacyWatchlist(from: legacyContext)
        reloadIndex()
    }

    // MARK: - Read

    func isInWatchlist(_ ref: MediaRef) -> Bool {
        refs.contains(ref)
    }

    // MARK: - Local-first writes

    func toggle(_ ref: MediaRef, metadata: WatchlistMetadata? = nil) {
        setWatchlist(!isInWatchlist(ref), for: ref, metadata: metadata)
    }

    func setWatchlist(
        _ inWatchlist: Bool,
        for ref: MediaRef,
        metadata: WatchlistMetadata? = nil
    ) {
        do {
            let existing = try item(for: ref)
            guard existing != nil || inWatchlist else { return }

            let item = existing ?? WatchlistItem(ref: ref, metadata: metadata)
            if existing == nil { modelContext.insert(item) }
            if let metadata { item.apply(metadata) }

            let stateChanged = item.isInWatchlist != inWatchlist
            item.isInWatchlist = inWatchlist
            item.updatedAt = Date()
            if stateChanged || existing == nil {
                item.pendingAction = inWatchlist ? .add : .remove
                if inWatchlist { item.dateAdded = Date() }
            }

            try modelContext.save()
            reloadIndex()
            if stateChanged || existing == nil { requestPendingPush() }
        } catch {
            lastSyncError = error.localizedDescription
        }
    }

    func remove(_ item: WatchlistItem) {
        setWatchlist(false, for: item.ref, metadata: WatchlistMetadata(item.snapshot))
    }

    /// Keeps a locally saved card fresh when its full detail page is loaded.
    func updateMetadata(_ detail: MediaDetail) {
        guard isInWatchlist(detail.ref) else { return }
        do {
            guard let item = try item(for: detail.ref) else { return }
            item.apply(WatchlistMetadata(detail))
            try modelContext.save()
            reloadIndex()
        } catch {
            lastSyncError = error.localizedDescription
        }
    }

    // MARK: - TMDB sync

    /// Flushes durable local mutations, then makes confirmed local rows match
    /// the connected account's complete movie and TV watchlists.
    func syncFromTMDB() async {
        guard isAuthenticated() else { return }
        guard !fullSyncRequested, !fullSyncInProgress else { return }
        fullSyncRequested = true
        await runRemoteWorkIfNeeded()
    }

    private func requestPendingPush() {
        guard isAuthenticated() else { return }
        pendingPushRequested = true
        Task { [weak self] in
            await self?.runRemoteWorkIfNeeded()
        }
    }

    /// Serializes change pushes and lifecycle pulls. If another request
    /// arrives during an await, the loop picks it up before going idle.
    private func runRemoteWorkIfNeeded() async {
        guard !isRunningRemoteWork, isAuthenticated() else { return }
        isRunningRemoteWork = true
        isSyncing = true
        lastSyncError = nil
        defer {
            isRunningRemoteWork = false
            isSyncing = false
        }

        while isAuthenticated() {
            let shouldPull = fullSyncRequested
            let shouldPush = pendingPushRequested || shouldPull
            guard shouldPush else { break }

            fullSyncRequested = false
            if shouldPull { fullSyncInProgress = true }
            pendingPushRequested = false
            await pushPendingChanges()
            if shouldPull {
                await pullFromTMDB()
                fullSyncInProgress = false
            }
        }
    }

    /// Successful writes remain pending until a subsequent list pull confirms
    /// them. That prevents a briefly stale TMDB list from undoing a local tap.
    private func pushPendingChanges() async {
        let pending: [WatchlistItem]
        do {
            pending = try allItems()
                .filter { $0.pendingAction != .none }
                .sorted { $0.updatedAt < $1.updatedAt }
        } catch {
            lastSyncError = error.localizedDescription
            return
        }

        for item in pending {
            guard isAuthenticated() else { return }
            let action = item.pendingAction
            guard action != .none else { continue }
            do {
                try await account.setWatchlist(action == .add, for: item.ref)
            } catch {
                lastSyncError = error.localizedDescription
            }
        }
    }

    private func pullFromTMDB() async {
        do {
            async let movies = account.watchlistItems(.movie)
            async let shows = account.watchlistItems(.tv)
            let remote = try await movies + shows
            try reconcile(with: remote)
        } catch {
            lastSyncError = error.localizedDescription
        }
    }

    /// Remote state is authoritative only for rows with no unconfirmed local
    /// mutation. Pending adds/removes win until TMDB's list reflects them.
    private func reconcile(with remoteItems: [TMDBMediaItem]) throws {
        let localItems = try allItems()
        var localByKey = Dictionary(
            uniqueKeysWithValues: localItems.map { ($0.storageKey, $0) }
        )
        let remoteKeys = Set(remoteItems.map { WatchlistItem.key(for: $0.ref) })

        for remote in remoteItems {
            let key = WatchlistItem.key(for: remote.ref)
            let local: WatchlistItem
            if let existing = localByKey[key] {
                local = existing
            } else {
                local = WatchlistItem(
                    ref: remote.ref,
                    metadata: WatchlistMetadata(remote),
                    pendingAction: .none
                )
                modelContext.insert(local)
                localByKey[key] = local
            }

            local.apply(WatchlistMetadata(remote))
            switch local.pendingAction {
            case .remove:
                // A local removal has not appeared on TMDB yet.
                local.isInWatchlist = false
            case .add, .none:
                local.isInWatchlist = true
                local.pendingAction = .none
            }
        }

        for local in localItems where !remoteKeys.contains(local.storageKey) {
            switch local.pendingAction {
            case .add:
                // Keep the local addition until TMDB confirms it.
                local.isInWatchlist = true
            case .remove, .none:
                // A pending removal is now confirmed; a confirmed row absent
                // remotely was removed outside Edendale.
                modelContext.delete(local)
            }
        }

        try modelContext.save()
        reloadIndex()
    }

    // MARK: - Persistence helpers

    private func item(for ref: MediaRef) throws -> WatchlistItem? {
        let key = WatchlistItem.key(for: ref)
        var descriptor = FetchDescriptor<WatchlistItem>(
            predicate: #Predicate { $0.storageKey == key }
        )
        descriptor.fetchLimit = 1
        return try modelContext.fetch(descriptor).first
    }

    private func allItems() throws -> [WatchlistItem] {
        try modelContext.fetch(FetchDescriptor<WatchlistItem>())
    }

    private func reloadIndex() {
        do {
            items = try allItems().sorted { $0.dateAdded > $1.dateAdded }
            refs = Set(items.filter(\.isInWatchlist).map(\.ref))
        } catch {
            lastSyncError = error.localizedDescription
        }
    }

    /// Moves watchlist flags written by older Edendale builds out of the
    /// CoreData/CloudKit account-state row and into the new local SwiftData
    /// store. The legacy attribute remains in the model for store migration.
    private func migrateLegacyWatchlist(from legacyContext: NSManagedObjectContext?) {
        guard let legacyContext else { return }
        let request = NSFetchRequest<CDUserMedia>(entityName: "CDUserMedia")
        request.predicate = NSPredicate(format: "inWatchlist == YES")

        do {
            let legacyItems = try legacyContext.fetch(request)
            guard !legacyItems.isEmpty else { return }
            var existingKeys = Set(try allItems().map(\.storageKey))

            for legacy in legacyItems {
                let type = TMDBMediaType(rawValue: legacy.mediaType) ?? .movie
                let ref = MediaRef(id: Int(legacy.tmdbId), mediaType: type)
                if existingKeys.insert(WatchlistItem.key(for: ref)).inserted {
                    modelContext.insert(
                        WatchlistItem(
                            ref: ref,
                            pendingAction: .add,
                            dateAdded: legacy.updatedAt ?? Date()
                        )
                    )
                }
            }
            try modelContext.save()

            for legacy in legacyItems { legacy.inWatchlist = false }
            if legacyContext.hasChanges { try legacyContext.save() }
        } catch {
            lastSyncError = error.localizedDescription
        }
    }
}

File: Edendale.Windows/Services/WatchlistStore.cs
// The watchlist's source of truth. Local mutations are saved first and mirrored
// to TMDB when an account is connected; a durable pending action survives a
// failed or interrupted request and is retried on the next sync. A full sync
// pulls the account's complete lists and lets the cloud win over confirmed
// local rows (pending taps still win until TMDB reflects them). Legacy watchlist
// flags written by older builds into user-media.json are migrated in on first
// run. This is a JSON-backed, WinUI-free service (like UserMediaStore), so the
// domain tests link it and inject a fake remote. Ports WatchlistStore.swift.

using System.Text.Json;
using Edendale.Windows.Core;
using Edendale.Windows.Models;

namespace Edendale.Windows.Services;

public sealed class WatchlistStore
{
    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
        PropertyNameCaseInsensitive = true,
        WriteIndented = true,
    };

    private readonly Dictionary<string, WatchlistRecord> _records = [];
    private readonly object _gate = new();
    private readonly IWatchlistRemote _remote;
    private readonly Func<bool> _isAuthenticated;
    private readonly string _storePath;
    private readonly string _legacyUserMediaPath;

    // Serialized remote work (coalesces overlapping push/pull requests).
    private readonly object _pumpGate = new();
    private bool _pumpRunning;
    private bool _pushRequested;
    private bool _pullRequested;

    /// <summary>Any change to the watchlist (local edit, migration, or sync).</summary>
    public event EventHandler? Changed;

    /// <param name="storePath">Watchlist JSON path; defaults to the app data file.</param>
    /// <param name="legacyUserMediaPath">Legacy user-media JSON to migrate from; defaults to the app data file.</param>
    public WatchlistStore(
        IWatchlistRemote remote,
        Func<bool> isAuthenticated,
        string? storePath = null,
        string? legacyUserMediaPath = null)
    {
        _remote = remote;
        _isAuthenticated = isAuthenticated;
        _storePath = storePath ?? AppPaths.WatchlistFile;
        _legacyUserMediaPath = legacyUserMediaPath ?? AppPaths.UserMediaFile;
        Load();
    }

    // ------------------------------------------------------------------
    // Read
    // ------------------------------------------------------------------

    /// <summary>Saved titles, newest first. Removed-but-unconfirmed rows are hidden.</summary>
    public IReadOnlyList<WatchlistRecord> Items
    {
        get
        {
            lock (_gate)
            {
                return [.. _records.Values
                    .Where(record => record.InWatchlist)
                    .OrderByDescending(record => record.DateAdded)];
            }
        }
    }

    public bool IsInWatchlist(MediaRef reference)
    {
        lock (_gate)
        {
            return _records.TryGetValue(WatchlistRecord.Key(reference), out var record)
                && record.InWatchlist;
        }
    }

    // ------------------------------------------------------------------
    // Local-first writes
    // ------------------------------------------------------------------

    public void Toggle(MediaRef reference, WatchlistMetadata? metadata = null) =>
        SetWatchlist(!IsInWatchlist(reference), reference, metadata);

    public void SetWatchlist(bool inWatchlist, MediaRef reference, WatchlistMetadata? metadata = null)
    {
        var changed = false;
        lock (_gate)
        {
            var key = WatchlistRecord.Key(reference);
            _records.TryGetValue(key, out var existing);
            // Nothing to do when removing a title that was never saved.
            if (existing is null && !inWatchlist) return;

            var record = existing ?? WatchlistRecord.Create(
                reference, metadata, inWatchlist: inWatchlist, pending: WatchlistPendingAction.None);
            if (existing is null) _records[key] = record;
            else if (metadata is not null) record.Apply(metadata);

            var stateChanged = record.InWatchlist != inWatchlist;
            record.InWatchlist = inWatchlist;
            record.UpdatedAt = NowMillis();
            if (stateChanged || existing is null)
            {
                record.PendingAction = inWatchlist ? WatchlistPendingAction.Add : WatchlistPendingAction.Remove;
                if (inWatchlist) record.DateAdded = NowMillis();
                changed = true;
            }
        }

        Save();
        RaiseChanged();
        if (changed) RequestPush();
    }

    public void Remove(WatchlistRecord record) => SetWatchlist(false, record.Ref);

    /// <summary>Keeps a saved card fresh when its full detail page loads.</summary>
    public void UpdateMetadata(MediaDetail detail)
    {
        if (!IsInWatchlist(detail.Ref)) return;
        lock (_gate)
        {
            if (!_records.TryGetValue(WatchlistRecord.Key(detail.Ref), out var record)) return;
            record.Apply(WatchlistMetadata.From(detail));
        }
        Save();
        RaiseChanged();
    }

    // ------------------------------------------------------------------
    // TMDB sync (cloud takes priority)
    // ------------------------------------------------------------------

    /// <summary>Flushes pending local mutations, then makes confirmed local rows
    /// match the connected account's complete movie and TV watchlists.</summary>
    public Task SyncFromTMDBAsync()
    {
        if (!_isAuthenticated()) return Task.CompletedTask;
        lock (_pumpGate) { _pullRequested = true; }
        return PumpAsync();
    }

    private void RequestPush()
    {
        if (!_isAuthenticated()) return;
        lock (_pumpGate) { _pushRequested = true; }
        _ = PumpAsync();
    }

    /// <summary>
    /// Serializes change pushes and full pulls. A request that arrives while a
    /// pump is running is observed by that pump before it goes idle, so no
    /// wake-up is lost. Push and pull never throw, so the loop only exits under
    /// the flag lock.
    /// </summary>
    private async Task PumpAsync()
    {
        lock (_pumpGate)
        {
            if (_pumpRunning) return;
            _pumpRunning = true;
        }

        try
        {
            while (_isAuthenticated())
            {
                bool pull, push;
                lock (_pumpGate)
                {
                    pull = _pullRequested;
                    push = _pushRequested || pull;
                    if (!push)
                    {
                        _pumpRunning = false;
                        return;
                    }
                    _pullRequested = false;
                    _pushRequested = false;
                }

                await PushPendingChangesAsync();
                if (pull) await PullFromTmdbAsync();
            }
        }
        finally
        {
            lock (_pumpGate) { _pumpRunning = false; }
        }
    }

    /// <summary>Successful writes stay pending until a subsequent pull confirms
    /// them; that stops a briefly stale TMDB list from undoing a local tap.</summary>
    private async Task PushPendingChangesAsync()
    {
        List<WatchlistRecord> pending;
        lock (_gate) pending = WatchlistReconciler.PendingPushOrder(_records.Values);

        foreach (var record in pending)
        {
            if (!_isAuthenticated()) return;
            var action = record.PendingAction;
            if (action == WatchlistPendingAction.None) continue;
            try
            {
                await _remote.SetWatchlistAsync(record.Ref, action == WatchlistPendingAction.Add);
            }
            catch
            {
                // Leave the pending action in place; a later pump retries it.
            }
        }
    }

    private async Task PullFromTmdbAsync()
    {
        try
        {
            var moviesTask = _remote.WatchlistItemsAsync("movie");
            var showsTask = _remote.WatchlistItemsAsync("tv");
            await Task.WhenAll(moviesTask, showsTask);

            var remote = new List<MediaItem>();
            remote.AddRange(await moviesTask);
            remote.AddRange(await showsTask);

            lock (_gate)
            {
                var reconciled = WatchlistReconciler.Reconcile([.. _records.Values], remote);
                _records.Clear();
                foreach (var record in reconciled) _records[record.StorageKey] = record;
            }
            Save();
            RaiseChanged();
        }
        catch
        {
            // Transient network/auth failure; the next sync retries.
        }
    }

    // ------------------------------------------------------------------
    // Persistence + migration
    // ------------------------------------------------------------------

    private void Load()
    {
        try
        {
            if (File.Exists(_storePath))
            {
                var stored = JsonSerializer.Deserialize<List<WatchlistRecord>>(
                    File.ReadAllText(_storePath), JsonOptions) ?? [];
                foreach (var record in stored) _records[record.StorageKey] = record;
                return;
            }

            // First run on this device: adopt any watchlist flags older builds
            // wrote into user-media.json, then write watchlist.json so the
            // migration runs exactly once.
            MigrateLegacyWatchlist();
            Save();
        }
        catch
        {
            // A corrupt store starts empty rather than blocking launch.
        }
    }

    private void MigrateLegacyWatchlist()
    {
        if (!File.Exists(_legacyUserMediaPath)) return;
        using var document = JsonDocument.Parse(File.ReadAllText(_legacyUserMediaPath));
        if (document.RootElement.ValueKind != JsonValueKind.Array) return;

        foreach (var element in document.RootElement.EnumerateArray())
        {
            if (element.Bool("watchlist") != true || element.Int("tmdbId") is not int id) continue;
            var reference = new MediaRef { Id = id, MediaType = element.String("mediaType") ?? "movie" };
            var key = WatchlistRecord.Key(reference);
            if (_records.ContainsKey(key)) continue;

            var metadata = new WatchlistMetadata
            {
                Title = element.String("title"),
                PosterPath = element.String("posterPath"),
                PosterUrl = element.String("posterUrl"),
            };
            _records[key] = WatchlistRecord.Create(
                reference,
                metadata,
                inWatchlist: true,
                pending: WatchlistPendingAction.Add,
                dateAdded: LegacyMillis(element, "watchlistUpdatedAt"));
        }
    }

    private static long LegacyMillis(JsonElement element, string name) =>
        element.Property(name) is { ValueKind: JsonValueKind.Number } value && value.TryGetInt64(out var millis)
            ? millis
            : NowMillis();

    private void Save()
    {
        try
        {
            List<WatchlistRecord> snapshot;
            lock (_gate) snapshot = [.. _records.Values];
            var temporary = _storePath + ".tmp";
            File.WriteAllText(temporary, JsonSerializer.Serialize(snapshot, JsonOptions));
            File.Move(temporary, _storePath, overwrite: true);
        }
        catch
        {
            // The watchlist is convenience data; a failed write must never crash the UI.
        }
    }

    private void RaiseChanged() => Changed?.Invoke(this, EventArgs.Empty);

    private static long NowMillis() => DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
}
